src/modules-load/modules-load.c

```diff
@@ -39,45 +88,249 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+                if (r < 0)
+                        return log_error_errno(r, "Failed to receive from module queue: %m");
+
+                if (r == 0) {
+                        log_debug("No more queued modules, terminate thread");
                         break;
+                }
 
-                if (isempty(line))
-                        continue;
-                if (strchr(COMMENTS, *line))
-                        continue;
+                r = module_load_and_warn(ctx, buffer, /* verbose = */ true);
+                if (r != -ENOENT)
+                        RET_GATHER(ret, r);
+        }
 
-                k = module_load_and_warn(ctx, line, true);
-                if (k == -ENOENT)
-                        continue;
-                RET_GATHER(r, k);
+        return ret;
+}
+
+static void *prober_thread(void *arg) {
+        int sock = PTR_TO_FD(arg);
+        return INT_TO_PTR(run_prober(sock));
+}
+
+static int create_worker_threads(
+                size_t n_threads,
+                void *arg,
+                pthread_t **ret_threads,
+                size_t *ret_n_threads) {
+        _cleanup_free_ pthread_t *new_threads = NULL;
+        size_t created_threads;
+        sigset_t ss, saved_ss;
+        int r;
+
+        assert(ret_threads);
+        assert(ret_n_threads);
+
+        if (n_threads == 0) {
+                *ret_n_threads = 0;
+                return 0;
+        }
+
+        /* Create worker threads with masked signals */
+        new_threads = new(pthread_t, n_threads);
+        if (!new_threads)
+                return log_oom();
+
+        /* No signals in worker threads. */
+        assert_se(sigfillset(&ss) >= 0);
+        r = pthread_sigmask(SIG_BLOCK, &ss, &saved_ss);
+        if (r != 0)
+                return log_error_errno(r, "Failed to mask signals for workers: %m");
+
+        for (created_threads = 0; created_threads < n_threads; ++created_threads) {
+                r = pthread_create(&new_threads[created_threads], NULL, prober_thread, arg);
+                if (r != 0) {
+                        log_error_errno(r, "Failed to create worker thread %zu: %m", created_threads);
+                        break;
+                }
         }
 
-        return r;
+        /* Restore the signal mask */
+        r = pthread_sigmask(SIG_SETMASK, &saved_ss, NULL);
+        if (r != 0)
+                log_error_errno(r, "Failed to restore signal mask: %m, ignoring");
+
+        *ret_threads = TAKE_PTR(new_threads);
+        *ret_n_threads = created_threads;
+
+        return 0;
+}
+
+static int destroy_worker_threads(pthread_t **threads, size_t n_threads) {
+        int ret = 0, r;
+
+        assert(threads);
+        assert(n_threads == 0 || *threads);
+
+        for (size_t i = 0; i < n_threads; ++i) {
+                void *p;
+                r = pthread_join((*threads)[i], &p);
+                if (r != 0)
+                        RET_GATHER(ret, log_error_errno(r, "Failed to join worker thread: %m"));
```

Comment:
as above, given that we proceed here after a failure please downgrade to warning, and say `…, proceeding anyway: %m` or so in the message